README.md
# Codeignter 3 Files Creator for vcode
[![Latest Release](https://vsmarketplacebadge.apphb.com/version-short/SyedMuradAliShah.codeigniter-3-files-creator.svg)](https://marketplace.visualstudio.com/items?itemName=SyedMuradAliShah.codeigniter-3-files-creator)
[![Installs](https://vsmarketplacebadge.apphb.com/installs-short/SyedMuradAliShah.codeigniter-3-files-creator.svg)](https://marketplace.visualstudio.com/items?itemName=SyedMuradAliShah.codeigniter-3-files-creator)
[![Rating](https://vsmarketplacebadge.apphb.com/rating-short/SyedMuradAliShah.codeigniter-3-files-creator.svg)](https://marketplace.visualstudio.com/items?itemName=SyedMuradAliShah.codeigniter-3-files-creator#review-details)
[![Fork](https://img.shields.io/github/forks/SyedMuradAliShah/codeigniter-3-files-creator.svg)](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator)
[![All Contributors](https://img.shields.io/badge/all%20contributors-1-blue)](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator/blob/master/CONTRIBUTORS.md#contributors)
[![Open Issues](https://img.shields.io/github/issues-raw/SyedMuradAliShah/codeigniter-3-files-creator.svg?style=flat)](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator/issues?q=is%3Aissue+is%3Aopen)
[![Closed Issues](https://img.shields.io/github/issues-closed-raw/SyedMuradAliShah/codeigniter-3-files-creator.svg?style=flat)](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator/issues?q=is%3Aissue+is%3Aclosed)
[![Donate](https://img.shields.io/liberapay/patrons/SyedMuradAliShah.svg?style=flat)](https://liberapay.com/SyedMuradAliShah/donate)


Creating files never been so easy, CI3 Files Creator bring you comands to create necessary with a few step.

**Features**
- Create hook
- Create template
- Create migration add table
- Create migration rename table
- Create migration add column to table
- Create migration rename column in table
- Create simple_auth
- Create simple_auth_api
- Create crud (include Model & Controller)
- Create model file
- Create controller file
- Create library file
- Create language file
- Create helper file
- Create htaccess file

### How to use?
Press `F1` then type `CI3` and you will see a bunch of commands.

If ask for folder?
- Give a name or Leave it empty. (If folder not found it will create new, else choose existing one.)

Asked for file name?
- Give file name

And that's all.

### How to use Template
1. Press `F1` and then `CI3 make:template` now choose your folder to if you want to make otherwise leave it empty, you can also use folder within folder, like `admin/dashboard` or `user/template/front_end` or `user/template/backend_end` etc.
2. Load template library `application\config\autoload.php` here.
3. Now in controller you have to call 
- `$this->TEMPLATE_FILE_NAME->load('VIEW_FILE_NAME', $data);`

For example if I set TEMPLATE_FILE_NAME as template and I have a view file name page.php
- `$this->template->load('page');`
- `$this->template->load('page', $data);`

Or you can use sub folder like this. 
- `$this->template->load('folder_name/page', $data);`

> NOTE: if you used a folder for template like `user_side` then your view files must be inside theh user_side directory!

And that's all.


### How to use Migration
1. enable migration `application\config\migration.php` here.
2. `migration_type` must be `timestamp`, if you set to Sequential migration won't work.

And that's all.

### How to use Migration Add Column to a table
1. All you need here to enter a valid format i.e `add_column_to_table` you can see following examples.
- You need to user `add_` at the start and then use `_to_` after column and before the table name. 
- Examples:
- `add_name_to_users`
- `add_first_name_to_users`
- `add_first_name_to_users_details`


### How to use Migration rename Column to a table
1. All you need here to enter a valid format i.e `rename_column_to_column2_in_table` you can see following examples.
- You need to user `rename_` at the start and then use `_to_` after old column and then use new column name before `_in_` and then use table name at last. 
- Examples:
- `rename_name_to_username_in_users`
- `rename_first_name_to_full_name_in_users`
- `rename_first_name_to_full_name_in_users_details`


Press `F1` then type bellow commands

| Command                            |                                                Function |
| :--------------------------------- | ------------------------------------------------------: |
| `CI3 make:hook`                    |                    It will create hook in hooks folder. |
| `CI3 make:template`                |              It will create template in library folder. |
| `CI3 make:migration:add_table`     |      It will create migration file in migration folder. |
| `CI3 make:migration:rename_table`  |      It will create migration file in migration folder. |
| `CI3 make:migration:add_column`    |   It will create migration for adding column to a table |
|                                    |                               file in migration folder. |
| `CI3 make:migration:rename_column` | It will create migration for renaming column in a table |
|                                    |                               file in migration folder. |
| `CI3 make:simple_auth`             |            It will create controller, model of complete |
|                                    |                                 login and registration. |
| `CI3 make:simple_auth_api`         |      It will create controller, model of complete login |
|                                    |           and registration and will return json format. |
| `CI3 make:crud`                    |   It will create controller, model with crud functions. |
| `CI3 make:model`                   |                              It will create model file. |
| `CI3 make:controller`              |                         It will create controller file. |
| `CI3 make:library`                 |          It will create library file in library folder. |
| `CI3 make:language`                |        It will create language file in language folder. |
| `CI3 make:helper`                  |            It will create helper file in helper folder. |
| `CI3 make:htaccess`                |                It will create `.htaccess` file in root. |



> This repository is public, your contribution will be appricated.


**You can support me by donating, My Bitcoin Address: 1GPLxZQwzAeWgERvxSe5WMmsMDGiURtQuS**

[![Donate](https://i.imgur.com/W6ggNR5.png)](http://bitcoin:1GPLxZQwzAeWgERvxSe5WMmsMDGiURtQuS)

[![Donate](https://i.imgur.com/2v7VgCu.png)](http://bitcoin:1GPLxZQwzAeWgERvxSe5WMmsMDGiURtQuS)

CHANGELOG.md
# Change Log
All changes will be updated here.

## [0.2.9] - 2022-07-13
- `.htaccess` improved
- `.htaccess` https and www and non-www version added
- `.htaccess` if https or www on, localhost will not use it.

## [0.2.8] - 2022-07-13
- Bug in blank `CRUD` fixed.

## [0.2.7] - 2022-07-08
- Creating view file is added
- `Template` will now point to folder pages
- In `CRUD` when you choose blank, it will create blank model & controller

## [0.2.6] - 2022-06-15
- Critical bug in `.htaccess` fixed

## [0.2.5] - 2022-06-11
- `.htaccess` rules updated

## [0.2.4] - 2022-03-07
- Some security added `.htaccess`

## [0.2.3] - 2022-02-16
- Critical error fixed in `Google Authenticator Library`

## [0.2.2] - 2022-02-16
- `CURD` minor bugs fixed  

## [0.2.1] - 2022-02-16
- `Google Authenticator Library` Added
- `CURD` fully-written and functional


## [0.2.0] - 2022-02-02
- Issue reported fixed [#4](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator/issues/4)

## [0.1.9] - 2022-01-23
- Issue reported fixed [#2](https://github.com/SyedMuradAliShah/codeigniter-3-files-creator/issues/2)

## [0.1.8] - 2021-12-12
- `Hook` added
- Some important functions added to helper function
- Default constructor added to library
- Open all files after creation.

## [0.1.7] - 2021-12-07
- Critical error fixed in `crud`
- `Language` fixed
- `Simple Auth` little bit improved

## [0.1.6] - 2021-12-07
- `Template` restricted to folder

## [0.1.5] - 2021-12-07
- Response text imporved in `template`

## [0.1.4] - 2021-12-07
- Functionality improved
- `Template` now accept folders
  
## [0.1.3] - 2021-12-02
- Functionality & helping text improved
- Bug in fixed
- `Migration` commands updated
- New `migration add column to table` added
- New `migration renew column in table` added

## [0.1.2] - 2021-08-05
- Minor bugs fix

## [0.1.1] - 2021-07-17
- Bug in `helper` fixed
- Bug in `template` fixed

## [0.1.0] - 2021-07-17
- Bug in `curd` fixed

## [0.0.9] - 2021-07-17
- Minor bug in `template` fixed

## [0.0.8] - 2021-07-17
- `Template` added
- Bug in `curd` fixed

## [0.0.7] - 2021-06-08
- Minor bugs fixed

## [0.0.6] - 2021-06-08
- Functionality improved

## [0.0.5] - 2021-04-15
- `Migration` added

## [0.0.4] - 2021-02-27
- `Simple Auth` Added
- `Simple Auth API` Added
- Folder selection and creation facility added
- `CRUD` Modified
- `Controllers` Modified
- `Models` Modified

## [0.0.3] - 2021-02-25
- Bug in `CRUD` fixed.
- Bug in `Helper` fixed.

## [0.0.2] - 2021-02-25
- `CRUD` function has been added

## [0.0.1] - 2021-02-24
- Initial release.


commands/mMigration.js
var capitalize = require('./functions');
module.exports = function (vscode, fs, path, pathdir) {
    vscode.window.showInputBox({
        prompt: "name of the table",
        placeHolder: "Enter the name of the table"
    }).then(function (val) {
        if (val.length == 0) {
            vscode.window.showErrorMessage("Table name is required.");

        } else {
            var d = new Date();
            var months = ["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12"];
            var FileTimeStamp = d.getFullYear() + months[d.getMonth()] + d.getDate() + d.getHours() + d.getMinutes() + d.getSeconds();

            var migrationDir = pathdir + "/application/migrations";
            var migrationFile = FileTimeStamp + '_' + val;
            var pathfile = path.join(migrationDir, migrationFile) + ".php";

            fs.access(pathfile, function (err) {
                if (!err) {
                    vscode.window.showWarningMessage("Table file name already exists!");

                } else {

                    if (!fs.existsSync(migrationDir)) {
                        fs.mkdirSync(migrationDir);
                        vscode.window.showInformationMessage("migrations folder created in applications.");
                    }

                    fs.open(pathfile, "w+", function (err, fd) {
                        if (err) throw err;
                        fs.writeFileSync(fd, `<?php

defined('BASEPATH') or exit('No direct script access allowed');

class Migration_` + capitalize.capitalize(val) + ` extends CI_Migration
{
    protected $tableName  = '` + val + `';

    public function up()
    {
        $this->dbforge->add_field(array(
            'id' => array(
                'type' => 'INT',
                'constraint' => 4,
                'unsigned' => TRUE,
                'auto_increment' => TRUE
            ),
            'username' => [
                'type' => 'VARCHAR',
                'constraint' => '30',
            ],
            'password' => [
                'type' => 'VARCHAR',
                'constraint' => '100',
            ],
            'status' => [
                'type'           => 'ENUM',
                'constraint'     => ['active', 'suspend'],
                'default'        => 'active',
            ]
        ));
        $this->dbforge->add_key('id', TRUE);
        $this->dbforge->add_field("updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP on update CURRENT_TIMESTAMP");
        $this->dbforge->add_field("created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP");

        // If you want to add a foriegn key.
        // role_id must be a column of this table, please add it above in the table. And make sure admin_roles table is added before this table. 
        // $this->dbforge->add_field('CONSTRAINT FOREIGN KEY (role_id) REFERENCES admin_roles(id) ON DELETE RESTRICT ON UPDATE CASCADE');

        $this->dbforge->create_table($this->tableName);

        //Inserting first row
        // $this->db->insert($this->tableName, [
        //     'username' => 'murad_ali',
        //     'phone' => '123-123-7834',
        //     'password' => password_hash('123456', PASSWORD_BCRYPT),
        // ]);
    }

    public function down()
    {
        $this->dbforge->drop_table($this->tableName);
    }
}


/* End of file ` + capitalize.capitalize(val) + `.php and path  /application/migrations/` + val + `.php */

`);
                        fs.close(fd);
                        var openPath = vscode.Uri.file(pathfile); //A request file path

                        vscode.workspace.openTextDocument(openPath).then(function (val) {
                            vscode.window.showTextDocument(val);

                        });

                    });
                    vscode.window.showInformationMessage('Migration table ' + migrationFile + ' has been created successfully! ');

                }


            });


        }

    });


}